frontend/workflows/projectSelector/src/helpers.tsx

```diff
@@ -0,0 +1,108 @@
+import * as React from "react";
+import _ from "lodash";
+
+import type { Action, Group, ProjectState, State } from "./types";
+
+const PROJECT_TYPE_URL = "type.googleapis.com/clutch.core.project.v1.Project";
+
+interface DependencyMappings {
+  upstreams?: { [dependency: string]: { [project: string]: boolean } };
+  downstreams?: { [dependency: string]: { [project: string]: boolean } };
+}
+
+const StateContext = React.createContext<State | undefined>(undefined);
+const useReducerState = () => {
+  return React.useContext(StateContext);
+};
+
+const DispatchContext = React.createContext<(action: Action) => void | undefined>(undefined);
+const useDispatch = () => {
+  return React.useContext(DispatchContext);
+};
+
+// TODO(perf): call with useMemo().
+const deriveSwitchStatus = (state: State, group: Group): boolean => {
+  return (
+    Object.keys(state[group]).length > 0 &&
+    Object.keys(state[group]).every(key => state[group][key].checked)
+  );
+};
+
+const updateGroupstate = (
+  state: State,
+  group: Group,
+  project: string,
+  projectState: ProjectState
+): State => {
+  const newState = { ...state };
+  if (project in newState[group]) {
+    // preserve the checked value if the project is already in the group
+    newState[group][project].checked = state[group][project].checked;
+    newState[group][project].custom = projectState.custom;
+  } else {
+    // we set the projectState to the default state passed in
+    newState[group][project] = projectState;
+  }
+
+  return newState;
+};
+
+const dependencyToProjects = (state: State, group: Group): DependencyMappings => {
+  const upstreamMap = {};
+  const downstreamMap = {};
+
+  const projects = Object.keys(state[group]);
+  projects.forEach(p => {
+    const { upstreams, downstreams } = state.projectData[p]?.dependencies || {};
+    upstreams?.[PROJECT_TYPE_URL]?.ids.forEach(u => {
+      if (!upstreamMap[u]) {
+        upstreamMap[u] = { [p]: true };
+      } else {
+        upstreamMap[u][p] = true;
+      }
+    });
+    downstreams?.[PROJECT_TYPE_URL]?.ids.forEach(d => {
+      if (!downstreamMap[d]) {
+        downstreamMap[d] = { [p]: true };
+      } else {
+        downstreamMap[d][p] = true;
+      }
+    });
+  });
+
+  return { upstreams: upstreamMap, downstreams: downstreamMap };
+};
+
+const exclusiveProjectDependencies = (
```

Comment:
@danielhochman  was thinking about this more and maybe we'll need a seperate helper for the hidden actions b/c the latter needs to know what is & isn't unchecked. ie if all the unchecked projects share the same dependency, they can be hidden but if any are still used by the checked projects, they can't.

Reply:
both can share the func that creates the mapping tho